apps/demos/Demos/Chat/AIAndChatbotIntegration/Angular/app/app.service.ts

```diff
@@ -0,0 +1,246 @@
+  }
+
+  async processMessageSending(message, event) {
+    this.toggleDisabledState(true, event);
+
+    this.messages.push({ role: 'user', content: message.text });
+    this.typingUsersSubject.next([this.assistant]);
+  
+    try {
+      const aiResponse = await this.getAIResponse(this.messages);
+
+      setTimeout(() => {
+        this.typingUsersSubject.next([]);
+        this.messages.push({ role: 'assistant', content: aiResponse });
+        this.renderAssistantMessage(aiResponse);
+      }, 200);
+    } catch {
+      this.typingUsersSubject.next([]);
+      this.messages.pop();
+      this.alertLimitReached();
+    } finally {
+      this.toggleDisabledState(false, event);
+    }
+  }
+
+  updateLastMessage(text = this.REGENERATION_TEXT) {
+    const items = this.dataSource.items();
+    const lastMessage = items.at(-1);
+  
+    this.dataSource.store().push([{
+      type: 'update',
+      key: lastMessage.id,
+      data: { text },
+    }]);
+  }
+
+  renderAssistantMessage(text: string) {
+    const message = {
+      id: Date.now(),
+      timestamp: new Date(),
+      author: this.assistant,
+      text,
+    };
+
+    this.dataSource.store().push([{ type: 'insert', data: message }]);
+  }
+
+  alertLimitReached() {
+    this.setAlerts([{
+      message: 'Request limit reached, try again in a minute.'
+    }]);
+  
+    setTimeout(() => {
+      this.setAlerts([]);
+    }, this.ALERT_TIMEOUT);
+  }
+
+  toggleDisabledState(disabled: boolean, event = undefined) {
+    this.isDisabled = disabled;
+
+    if (disabled) {
+      event?.target.blur();
+    } else {
+      event?.target.focus();
+    }
+  };
+
+  setAlerts(alerts: Alert[]) {
+    this.alerts = alerts;
+    this.alertsSubject.next(alerts);
+  }
+
+  async regenerate() {
+    this.toggleDisabledState(true);
+
+    try {
+      const aiResponse = await this.getAIResponse(this.messages.slice(0, -1));
+
+      this.updateLastMessage(aiResponse);
+      this.messages.at(-1).content = aiResponse;
+    } catch {
+      this.updateLastMessage(this.messages.at(-1).content);
+      this.alertLimitReached();
+    } finally {
+      this.toggleDisabledState(false);
+    }
+  }
+
+  convertToHtml(value: string) {
+    const result = unified()
+        .use(remarkParse)
+        .use(remarkRehype)
+        .use(rehypeStringify)
+        .processSync(value)
+        .toString();
+
+    return result;
+}
+  
+  onMessageEntered({ message, event }: MessageEnteredEvent) {
+    this.dataSource.store().push([{ type: 'insert', data: { id: Date.now(), ...message } }]);
+
+    if (!this.alerts.length) {
+      this.processMessageSending(message, event);
+    }
+  }
+
+  onCopyButtonClick(message: Message) {
+    navigator.clipboard?.writeText(message.text);
+    this.copyButtonIcon = 'check';
+
+    setTimeout(() => {
+        this.copyButtonIcon = 'copy';
+    }, 2500);
+}
+
+  onRegenerateButtonClick() {
+    this.updateLastMessage();
+    this.regenerate();
+  }
```

Comment:
I think too much stuff is in the service now. Everything related to view (dom / UI) should be in the component. Service is for data/utils/API calls.

Reply:
[995d5d8a2f2d789fb73b106bb05189b52e6568df]